src/lib/eco-router/hooks.ts

```diff
@@ -0,0 +1,165 @@
+import { Trade, CurrencyAmount, Currency, Percent } from '@swapr/sdk'
+import { useState, useEffect } from 'react'
+// Eco Router modules
+import { getExactIn, getExactOut } from './api'
+// Web3 hooks
+import { useActiveWeb3React } from '../../hooks'
+import { useIsMultihop } from '../../state/user/hooks'
+
+// Types
+// eslint-disable-next-line
+import type { EcoRouterHookResults } from './types'
+
+/**
+ * Eco Router hook
+ * @param currencyAmountIn Input currency amount
+ * @param currencyOut Output currency
+ * @returns {EcoRouterHookResults} List of trade sources
+ */
+export function useEcoRouterExactIn(currencyAmountIn?: CurrencyAmount, currencyOut?: Currency): EcoRouterHookResults {
+  const [loading, setLoading] = useState(true)
+  // All trades including Curve, Unsiwap V2 and CowSwap
+  const [trades, setTrades] = useState<Trade[]>([])
+  // Errors
+  const [errors, setErrors] = useState<any[]>([])
+  // Chain Id
+  const { chainId, account, library } = useActiveWeb3React()
+  // Uniswap V2 Trade option: using multi-hop option
+  const uniswapV2IsMultihop = useIsMultihop()
+  // Used to trigger computing trade route
+  const currencyInAndOutAdddress = `${currencyOut?.address}-${currencyAmountIn?.currency.address}`
+
+  useEffect(() => {
+    let isCancelled = false
+
+    // Early exit and clean state if necessary
+    if (!currencyAmountIn || !currencyAmountIn.currency || !currencyOut || !chainId) {
+      setTrades([])
+      setLoading(false)
+      return
+    }
+
+    // Reset state
+    setTrades([])
+    setLoading(true)
+
+    getExactIn(
+      {
+        currencyAmountIn,
+        currencyOut,
+        maximumSlippage: new Percent('3', '100'),
+        receiver: account ?? undefined,
+      },
+      {
+        uniswapV2: {
+          useMultihops: uniswapV2IsMultihop,
+        },
+      },
+      library
+    )
+      .then(newTrades => {
+        // Only update this invokation is not cancelled
+        if (!isCancelled) {
+          setTrades(newTrades.trades)
+          setErrors(newTrades.errors)
+        }
+      })
+      .catch(error => setErrors([error]))
+      .finally(() => setLoading(false))
+
+    return () => {
+      isCancelled = true
+    }
+    // eslint-disable-next-line
+  }, [
+    chainId,
+    uniswapV2IsMultihop,
+    // eslint-disable-next-line
+    currencyAmountIn?.toSignificant(),
```

Comment:
Why are we doing this? Should we use a `useRef` in this case?

Reply:
`currencyAmountIn` is a new object every time, causing infinite re-renders. Comparing the sting value is prevents this.